src/Gui/Inventor/Draggers/Gizmo.cpp

```diff
@@ -0,0 +1,697 @@
+    FC_ADD_CATALOG_ENTRY(toggleSwitch, SoToggleSwitch, annotation);
+    FC_ADD_CATALOG_ENTRY(geometry, SoSeparator, toggleSwitch);
+
+    SO_KIT_INIT_INSTANCE();
+
+    SO_KIT_ADD_FIELD(visible, (1));
+
+    auto pickStyle = SO_GET_ANY_PART(this, "pickStyle", SoPickStyle);
+    pickStyle->style = SoPickStyle::SHAPE_ON_TOP;
+
+    auto toggleSwitch = SO_GET_ANY_PART(this, "toggleSwitch", SoToggleSwitch);
+    toggleSwitch->on.connectFrom(&visible);
+
+    setPart("geometry", new SoSeparator);
+
+    cameraSensor.setFunction(&Gizmos::cameraChangeCallback);
+    cameraSensor.setData(this);
+
+    cameraPositionSensor.setData(this);
+    cameraPositionSensor.setFunction(cameraPositionChangeCallback);
+}
+
+Gizmos::~Gizmos()
+{
+    cameraSensor.setData(nullptr);
+    cameraSensor.detach();
+
+    cameraPositionSensor.setData(nullptr);
+    cameraPositionSensor.detach();
+
+    uninitGizmos();
+}
+
+void Gizmos::initGizmos()
+{
+    auto geometry = SO_GET_ANY_PART(this, "geometry", SoSeparator);
+    for (auto gizmo: gizmos) {
+        geometry->addChild(gizmo->initDragger());
+    }
+}
+
+void Gizmos::uninitGizmos()
+{
+    for (auto gizmo: gizmos) {
+        gizmo->uninitDragger();
+        delete gizmo;
+    }
+    gizmos.clear();
+}
+
+void Gizmos::addGizmo(Gizmo* gizmo)
+{
+    assert(std::ranges::find(gizmos, gizmo) == gizmos.end() && "this gizmo is already added!");
+    gizmos.push_back(gizmo);
+}
+
+void Gizmos::attachViewer(Gui::View3DInventorViewer* viewer, Base::Placement &origin)
+{
+    if (viewer) {
+        auto mat = origin.toMatrix();
+
+        viewer->getDocument()->setEditingTransform(mat);
+        So3DAnnotation* annotation = SO_GET_ANY_PART(this, "annotation", So3DAnnotation);
+        viewer->setupEditingRoot(annotation, &mat);
+    }
+}
+
+void Gizmos::setUpAutoScale(SoCamera* cameraIn)
+{
+    if (cameraIn->getTypeId() == SoOrthographicCamera::getClassTypeId()) {
+        auto localCamera = dynamic_cast<SoOrthographicCamera*>(cameraIn);
+        assert(localCamera);
+        cameraSensor.attach(&localCamera->height);
+        cameraPositionSensor.attach(&localCamera->orientation);
+        calculateScaleAndOrientation();
+
+    }
+    else if (cameraIn->getTypeId() == SoPerspectiveCamera::getClassTypeId()) {
+        auto localCamera = dynamic_cast<SoPerspectiveCamera*>(cameraIn);
+        assert(localCamera);
+        cameraSensor.attach(&localCamera->position);
+        cameraPositionSensor.attach(&localCamera->orientation);
+        calculateScaleAndOrientation();
+    }
+}
+
+void Gizmos::calculateScaleAndOrientation()
+{
+    if (cameraSensor.getAttachedField()) {
+        cameraChangeCallback(this, nullptr);
+        cameraPositionChangeCallback(this, nullptr);
+    }
+}
+
+void Gizmos::cameraChangeCallback(void* data, SoSensor*)
+{
+    assert(data);
+    auto sudoThis = static_cast<Gizmos*>(data);
+
+    SoField* field = sudoThis->cameraSensor.getAttachedField();
+    if (field) {
```

Comment:
Moving variable initialization to if condition ensures that it cannot be used outside of if.
```suggestion
    if (SoField* field = sudoThis->cameraSensor.getAttachedField()) {
```

Or you can do early exit.